controller/engine_controller.go

```diff
@@ -501,15 +501,38 @@ func (ec *EngineController) DeleteInstance(obj interface{}) error {
 		}
 	}
 
-	if im.Status.CurrentState != longhorn.InstanceManagerStateRunning {
-		return nil
+	v, err := ec.ds.GetVolumeRO(e.Spec.VolumeName)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+	}
+
+	isRWXVolume := false
+	if v != nil && v.Spec.AccessMode == longhorn.AccessModeReadWriteMany && !v.Spec.Migratable {
+		isRWXVolume = true
 	}
 
+	// For a RWX volume, the node down, for example, caused by kubelet restart, leads to share-manager pod deletion/recreation
+	// and volume detachment/attachment.
+	// Then, the newly created share-manager pod blindly mounts the longhorn volume inside /dev/longhorn/<pvc-name> and exports it.
+	// To avoid mounting a dead and orphaned volume, try to clean up the engine process as well as the orphaned iscsi device
+	// regardless of the instance-manager status.
+	if !isRWXVolume {
+		if im.Status.CurrentState != longhorn.InstanceManagerStateRunning {
+			logrus.Infof("Skip deleting engine %v since instance manager is in %v state", e.Name, im.Status.CurrentState)
+			return nil
+		}
+	}
+
+	logrus.Infof("Deleting engine instance %v", e.Name)
```

Comment:
One community user intentionally shut down a node where one RWX volume's engine instance is running on.
The RWX volume is stuck at detaching, because the engine instance is not able to be deleted.

Should we add a maximum tries for the stuck detaching?
cc @shuo-wu @innobead

Reply:
IIRC, users can detach a node attached to a down node. Why can't this engine be stopped?

Reply:
Previous, if the IM is not in running state, the `deleteInstance` is skipped directly. But in this case, we delete the engine instance for the RWX volume. If failed, retry again and again. So, the engine cannot be stopped.

Reply:
Got it. See the updated comment.

Reply:
> can the SM pod somehow understand if a volume is valid before mounting

No, the check logics are in share manager pod. It only see a `device`.
The `stat` can be executed and return correct information.  Then, fail to check the filesystem type (this is expected for a new volume) and do `mkfs`. It will encounter IO error when running `mkfs`.

Reply:
I see. Do you prefer to add a retry/backoff for this? Typically it requires a new field in engine CR... Here we may not be able to use an in-memory field.

Reply:
Do you mean retry `deleteInstance`?

Reply:
How about doing once? For kubelet restart, the underlay network and the IM pod are available. The deletion should succeed. If failed, the node can be regarded as a real shutdown.

Reply:
> Should we add a maximum tries for the stuck detaching?

Sorry... I was thinking you want to add a retry...

How about the network disconnection case? Will there be any leftovers/risks after the network recovery?

Reply:
Sounds good. This function (or the upper caller) will try its best to delete the process during cleanup but it cannot guarantee.

Is the fix PR already created?